File: src/main/scala/li/cil/oc/integration/util/Wrench.scala
package li.cil.oc.integration.util

import java.lang.reflect.Method

import li.cil.oc.common.IMC
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.item.ItemStack
import net.minecraft.util.Hand
import net.minecraft.util.math.BlockPos

import scala.collection.mutable

object Wrench {
  private val usages = mutable.LinkedHashSet.empty[Method]
  private val checks = mutable.LinkedHashSet.empty[Method]

  def addUsage(wrench: Method): Unit = usages += wrench

  def addCheck(checker: Method): Unit = checks += checker

  def isWrench(stack: ItemStack): Boolean = !stack.isEmpty && checks.exists(IMC.tryInvokeStatic(_, stack)(false))

  def holdsApplicableWrench(player: PlayerEntity, position: BlockPos): Boolean =
    !player.getItemInHand(Hand.MAIN_HAND).isEmpty && usages.exists(IMC.tryInvokeStatic(_, player, position, java.lang.Boolean.FALSE)(false))

  def wrenchUsed(player: PlayerEntity, position: BlockPos): Unit =
    if (!player.getItemInHand(Hand.MAIN_HAND).isEmpty) usages.foreach(IMC.tryInvokeStaticVoid(_, player, position, java.lang.Boolean.TRUE))
}

File: src/main/scala/li/cil/oc/integration/util/Wrench.java
package li.cil.oc.integration.util;

import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.ItemStack;
import net.minecraft.util.EnumHand;
import net.minecraft.util.math.BlockPos;

import javax.annotation.Nullable;
import java.util.ArrayList;
import java.util.List;

public final class Wrench {
    @FunctionalInterface
    public interface WrenchValidator {
        boolean isWrench(final ItemStack stack);
    }

    @FunctionalInterface
    public interface WrenchConsumer {
        boolean useWrench(final ItemStack stack, final BlockPos pos, final boolean simulate);
    }

    private static final List<WrenchValidator> validators = new ArrayList<>();
    private static final List<WrenchConsumer> consumers = new ArrayList<>();

    public static void addUsage(@Nullable final WrenchConsumer value) {
        if (value == null) {
            return;
        }

        if (!consumers.contains(value)) {
            consumers.add(value);
        }
    }

    public static void addValidator(@Nullable final WrenchValidator value) {
        if (value == null) {
            return;
        }

        if (!validators.contains(value)) {
            validators.add(value);
        }
    }

    public static boolean isWrench(final ItemStack stack) {
        if (stack.isEmpty()) {
            return false;
        }

        for (final WrenchValidator validator : validators) {
            if (validator.isWrench(stack)) {
                return true;
            }
        }

        return false;
    }

    public static boolean holdsApplicableWrench(final EntityPlayer player, final EnumHand hand, final BlockPos pos) {
        final ItemStack heldItem = player.getHeldItem(hand);
        if (heldItem.isEmpty()) {
            return false;
        }

        for (final WrenchConsumer consumer : consumers) {
            if (consumer.useWrench(heldItem, pos, true)) {
                return true;
            }
        }

        return false;
    }

    public static void wrenchUsed(final EntityPlayer player, final EnumHand hand, final BlockPos pos) {
        final ItemStack heldItem = player.getHeldItem(hand);
        if (heldItem.isEmpty()) {
            return;
        }

        for (final WrenchConsumer consumer : consumers) {
            consumer.useWrench(heldItem, pos, false);
        }
    }
}
